# Knoll et al., Frontiers in Immunology 2023: Identification of drug candidates targeting monocyte reprogramming in people living with HIV

<p align="center">
  <img src="https://github.com/knollr/HIV_pilot/blob/main/Graphical_Abstract_HIV.jpg" width="500" />
</p>

<p align="center" >
 Graphical Abstract generated with Biorender.com
</p>

<p align="justify">
  
**Introduction**: People living with HIV (PLHIV) are characterized by functional reprogramming of innate immune cells even after long-term antiretroviral therapy (ART). In order to assess technical feasibility of omics technologies for application to larger cohorts, we compared multiple omics data layers.

**Methods**: Bulk and single-cell transcriptomics, flow cytometry, proteomics, chromatin landscape analysis by ATAC-seq as well as ex vivo drug stimulation were performed in a small number of blood samples derived from PLHIV and healthy controls from the 200-HIV cohort study.

**Results**: Single-cell RNA-seq analysis revealed that most immune cells in peripheral blood of PLHIV are altered in their transcriptomes and that a specific functional monocyte state previously described in acute HIV infection is still existing in PLHIV while other monocyte cell states are only occurring acute infection. Further, a reverse transcriptome approach on a rather small number of PLHIV was sufficient to identify drug candidates for reversing the transcriptional phenotype of monocytes in PLHIV.

**Discussion**: These scientific findings and technological advancements for clinical application of single-cell transcriptomics form the basis for the larger 2000-HIV multicenter cohort study on PLHIV, for which a combination of bulk and single-cell transcriptomics will be included as the leading technology to determine disease endotypes in PLHIV and to predict disease trajectories and outcomes.

https://doi.org/10.3389/fimmu.2023.1275136
</p>

# Code and objects

This repository stores all code and data objects necessary to perform the analyses of this publication.
